core/image/jpeg/jpeg.odin

```diff
@@ -0,0 +1,927 @@
+										mcu[i * BLOCK_SIZE + 5] = cast(i16)(b2 - b5)
+										mcu[i * BLOCK_SIZE + 6] = cast(i16)(b1 - b6)
+										mcu[i * BLOCK_SIZE + 7] = cast(i16)(b0 - b7)
+									}
+								}
+							}
+						}
+
+						// Convert the YCbCr pixel data to RGB
+						cbcr_blk := &blocks[y * block_width + x]
+						for v := color_components[.Y].v_sampling_factor - 1; v >= 0; v -= 1 {
+							for h := color_components[.Y].h_sampling_factor - 1; h >= 0; h -= 1 {
+								y_blk := &blocks[(y + v) * block_width + (x + h)]
+								for j := BLOCK_SIZE - 1; j >= 0; j -= 1 {
+									for k := BLOCK_SIZE - 1; k >= 0; k -= 1 {
+										i := j * BLOCK_SIZE + k
+										cbcrPixelRow := j / color_components[.Y].v_sampling_factor + 4 * v
+										cbcrPixelColumn := k / color_components[.Y].h_sampling_factor + 4 * h
+										cbcrPixel := cbcrPixelRow * BLOCK_SIZE + cbcrPixelColumn
+
+										r := cast(i16)math.clamp(cast(f32)y_blk[.Y][i] + 1.402 * cast(f32)cbcr_blk[.Cr][cbcrPixel] + 128, 0, 255)
+										g := cast(i16)math.clamp(cast(f32)y_blk[.Y][i] - 0.344 * cast(f32)cbcr_blk[.Cb][cbcrPixel] - 0.714 * cast(f32)cbcr_blk[.Cr][cbcrPixel] + 128, 0, 255)
+										b := cast(i16)math.clamp(cast(f32)y_blk[.Y][i] + 1.772 * cast(f32)cbcr_blk[.Cb][cbcrPixel] + 128, 0, 255)
+
+										y_blk[.Y][i] = r
+										y_blk[.Cb][i] = g
+										y_blk[.Cr][i] = b
+									}
+								}
+							}
+						}
+					}
+				}
+
+				if resize(&img.pixels.buf, img.width * img.height * img.channels) != nil {
+					return img, .Unable_To_Allocate_Or_Resize
+				}
+
+				out := mem.slice_data_cast([]image.RGB_Pixel, img.pixels.buf[:])
+				for y in 0..<img.height {
+					mcu_row := y / BLOCK_SIZE
+					pixel_row := y % BLOCK_SIZE
+					for x in 0..<img.width {
+						mcu_col := x / BLOCK_SIZE
+						pixel_col := x % BLOCK_SIZE
+						mcu_idx := mcu_row * block_width + mcu_col
+						pixel_idx := pixel_row * BLOCK_SIZE + pixel_col
+
+						if img.channels == 3 {
+							out[y * img.width + x] = {
+								cast(byte)blocks[mcu_idx][.Y][pixel_idx],
+								cast(byte)blocks[mcu_idx][.Cb][pixel_idx],
+								cast(byte)blocks[mcu_idx][.Cr][pixel_idx],
+							}
+						} else {
+							img.pixels.buf[y * img.width + x] = cast(byte)blocks[mcu_idx][.Y][pixel_idx]
+						}
+					}
+				}
+
+				expect_EOI = true
+			case .TEM:
+				// TEM doesn't have a length, continue to next marker
+			case:
+				length := (compress.read_data(ctx, u16be) or_return) - 2
+				fmt.printfln("Unhandled marker: %v. Skipping %v bytes", marker, length)
+				compress.read_slice_from_memory(ctx, cast(int)length) or_return
+			}
+		}
+	}
+
+	return
+}
+
+destroy :: proc(img: ^Image) {
+	if img == nil {
+		return
+	}
+
+	bytes.buffer_destroy(&img.pixels)
+
+	if v, ok := img.metadata.(^image.JPEG_Info); ok {
+		if jfxx, jfxx_ok := v.jfxx_app0.?; jfxx_ok {
+			delete(jfxx.thumbnail)
+		}
+		if jfif, jfif_ok := v.jfif_app0.?; jfif_ok {
+			delete(jfif.thumbnail)
+		}
```

Comment:
Not strictly necessary to check if there is a jfxx or jfif part, calling `delete` on both thumbnails unconditionally is fine here since `delete(nil)` is a no-op.

Reply:
How would calling `delete` on the thumbnails directly work? `jfif_app0` is a `Maybe(JFIF_APP0)` and doing `delete(v.jfif_app0.?.thumbnail)` would result in an `Invalid type assertion` crash when `jfif_app0` is nil